init.lua

```diff
@@ -118,6 +118,30 @@ function HC:collisions(shape)
 	return candidates
 end
 
+function HC:raycast(x, y, dx, dy, range)
+	local dxr, dyr = dx * range, dy * range
+	local bbox = { x + dxr , y + dyr, x, y }
+	local candidates = self._hash:inSameCells(unpack(bbox))
+
+	for col in pairs(candidates) do
+		local rparams = col:intersectionsWithRay(x, y, dx, dy)
+		if #rparams > 0 then
+			for i, rparam in pairs(rparams) do
+				if rparam < 0 or rparam > range then
+					rawset(rparams, i, nil)
```

Comment:
Why use rawset?

Reply:
I found out that for many places we are using `rawset`, So I've also used it to keep some consistency between functions.

Reply:
IIRC the idea was that `rawset(tbl, key, val)` should be faster than `tbl[key] = val`, because it sidesteps `__newindex`. But that does not seem to be actually true, neither for vanilla lua nor for luajit:

```lua
local function bench(f)
  local t, N = {}, 1e6
  for _ = 1,10 do f(t, math.random()) end

  local tic = os.clock()
  for _ = 1, N do f(t, math.random()) end
  return os.clock() - tic
end

local t_index = bench(function(t, i) t[i] = true end)
local t_rawset = bench(function(t, i) rawset(t, i, true) end)

print('t_index:', t_index)
print('t_rawset:', t_rawset)
print('ratio:', t_rawset / t_index)
```

Vanilla:
```
t_index:	0.307589
t_rawset:	0.335716
ratio:	1.0914434521391
```

luajit:
```
t_index:	0.161559
t_rawset:	0.173773
ratio:	1.0756008640806
```

So, `rawset` is about 5-10% *slower* than using table access and it might be worth changing all `rawset`s except where its actually needed.

Reply:
Thanks for doing the benchmark, The results are interesting, I didn't expect that rawset could be slower, but I'm not sure why that's the case, Maybe rawset skipped some optimization steps? It can be the case in LuaJit but vanilla JS is also slower, So there is more to it.
In the future, I will try to change every rawset to index access and benchmark in on the project itself, If results show some major improvement then maybe we should ship it in the next version.